File: server/services/documentService.js
// server/services/documentService.js
// Mock document analysis: recognize doc types, extract fields, check against
// scheme requirements, and identify missing documents.

import { documentTypes, getDocumentRequirements } from '../data/documents.js';
import { getSchemeById } from './schemeService.js';

// ---------- Mock OCR / classification ----------
/**
 * Recognize a document type from a filename + (mock) content hint.
 * INTEGRATION POINT: Replace with real OCR/ML document classifier.
 */
export function recognizeDocumentType(fileName = '', hint = '') {
  const text = `${fileName} ${hint}`.toLowerCase();
  const keywordMap = {
    aadhaar: ['aadhaar', 'uidai', 'आधार'],
    pan: ['pan', 'permanent account'],
    'income-certificate': ['income', 'आय', 'income certificate'],
    'student-id': ['student', 'college', 'school', 'roll'],
    'bank-passbook': ['passbook', 'bank', 'account'],
    'ration-card': ['ration', 'राशन'],
    'land-proof': ['khata', 'land', 'patta', 'खसरा', 'जमीन'],
    'age-proof': ['birth', 'age', 'जन्म'],
    'birth-certificate': ['birth', 'जन्म'],
    'project-report': ['project', 'business plan', 'परियोजना'],
    'incorporation-cert': ['incorporation', 'cin', 'company'],
    'business-plan': ['business plan', 'funding'],
    'maternal-card': ['maternal', 'anc', 'गर्भ'],
    'sowing-certificate': ['sowing', 'crop', 'बुवाई'],
  };

  for (const [type, keys] of Object.entries(keywordMap)) {
    if (keys.some((k) => text.includes(k))) return type;
  }
  return 'unknown';
}

/**
 * Mock field extraction for a recognized document type.
 * INTEGRATION POINT: Replace with real OCR text parsing.
 */
export function extractFields(type, rawContent = {}) {
  const fields = documentTypes[type]?.fields || [];
  const extracted = {};
  fields.forEach((f) => {
    extracted[f] = rawContent[f] != null ? rawContent[f] : `[extracted ${f}]`;
  });
  return extracted;
}

/**
 * Analyze an uploaded document (mock). Returns recognized type, fields, and
 * confidence. Accepts a `file` (filename) and optional `content` blob.
 */
export function analyzeDocument({ fileName, content = {}, hint } = {}) {
  const type = recognizeDocumentType(fileName, hint);
  const extractedFields = type === 'unknown' ? {} : extractFields(type, content);
  return {
    type,
    label: documentTypes[type]?.label || 'Unknown Document',
    confidence: type === 'unknown' ? 0.4 : 0.92,
    extractedFields,
    fileName: fileName || 'uploaded-file',
    analyzedAt: new Date().toISOString(),
  };
}

/**
 * Check a user's uploaded documents against a scheme's requirements.
 * @returns {object} { schemeId, satisfied[], missing[] }
 */
export function checkDocumentsForScheme(schemeId, userDocuments = []) {
  const scheme = getSchemeById(schemeId);
  if (!scheme) throw new Error('Scheme not found');

  const required = scheme.requiredDocuments || [];
  const haveTypes = new Set(userDocuments.map((d) => d.type));
  const satisfied = [];
  const missing = [];

  required.forEach((docKey) => {
    const meta = documentTypes[docKey] || { label: docKey, howToGet: '', description: '' };
    if (haveTypes.has(docKey)) {
      satisfied.push({ key: docKey, label: meta.label });
    } else {
      missing.push({
        key: docKey,
        label: meta.label,
        description: meta.description,
        howToGet: meta.howToGet,
      });
    }
  });

  return { schemeId, schemeName: scheme.name, satisfied, missing };
}

/**
 * Get missing documents for a scheme given the user's uploaded docs.
 */
export function getMissingDocuments(schemeId, userDocuments = []) {
  const result = checkDocumentsForScheme(schemeId, userDocuments);
  return {
    schemeId,
    schemeName: result.schemeName,
    missing: result.missing,
    satisfied: result.satisfied,
    hasAll: result.missing.length === 0,
    guidance:
      result.missing.length === 0
        ? 'सभी आवश्यक दस्तावेज़ उपलब्ध हैं। आप आवेदन कर सकते हैं।'
        : 'कुछ दस्तावेज़ अभी लापता हैं। नीचे दिए गए तरीके से प्राप्त करें।',
  };
}

export { getDocumentRequirements };

export default {
  recognizeDocumentType,
  extractFields,
  analyzeDocument,
  checkDocumentsForScheme,
  getMissingDocuments,
  getDocumentRequirements,
};

File: ai-service/services/document_service.py
# services/document_service.py
# Document analysis via Groq vision + cross-doc name consistency checks.

from datetime import datetime, timezone
from data.documents import DOCUMENT_TYPES, SAMPLE_UPLOADED_DOCUMENTS
from services.scheme_service import get_scheme_by_id
from services.groq_ocr_service import check_cross_doc_names  # re-export for routes


# ── Scheme document matching ──────────────────────────────────────────────────

def check_documents_for_scheme(scheme_id: str, user_documents: list) -> dict:
    scheme = get_scheme_by_id(scheme_id)
    if not scheme:
        raise ValueError("Scheme not found")

    required   = scheme.get("requiredDocuments", [])
    have_types = {d.get("type") for d in user_documents}
    satisfied, missing = [], []

    for doc_key in required:
        meta = DOCUMENT_TYPES.get(doc_key, {"label": doc_key, "howToGet": "", "description": ""})
        if doc_key in have_types:
            matched = next((d for d in user_documents if d.get("type") == doc_key), {})
            satisfied.append({
                "key":             doc_key,
                "label":           meta.get("label", doc_key),
                "extractedFields": matched.get("extractedFields", {}),
                "warnings":        matched.get("warnings", []),
                "status":          matched.get("status", "verified"),
            })
        else:
            missing.append({
                "key":         doc_key,
                "label":       meta.get("label", doc_key),
                "description": meta.get("description", ""),
                "howToGet":    meta.get("howToGet", ""),
            })

    name_consistency = check_cross_doc_names(user_documents)

    return {
        "schemeId":        scheme_id,
        "schemeName":      scheme["name"],
        "satisfied":       satisfied,
        "missing":         missing,
        "nameConsistency": name_consistency,
    }


def get_missing_documents(scheme_id: str, user_documents: list) -> dict:
    result = check_documents_for_scheme(scheme_id, user_documents)
    return {
        **result,
        "hasAll": len(result["missing"]) == 0,
        "guidance": (
            "सभी आवश्यक दस्तावेज़ उपलब्ध हैं। आप आवेदन कर सकते हैं।"
            if len(result["missing"]) == 0
            else "कुछ दस्तावेज़ अभी लापता हैं। नीचे दिए गए तरीके से प्राप्त करें।"
        ),
    }


# ── Keyword-only compatibility path for the /upload JSON endpoint ─────────────

def _keyword_classify(file_name: str, hint: str) -> str:
    KEYWORD_MAP = {
        "aadhaar":            ["aadhaar", "aadhar", "uidai", "आधार"],
        "pan":                ["pan", "permanent account"],
        "income-certificate": ["income", "आय"],
        "student-id":         ["student", "college", "roll"],
        "bank-passbook":      ["passbook", "bank", "account"],
        "ration-card":        ["ration", "राशन"],
        "land-proof":         ["khata", "land", "patta", "खसरा"],
        "age-proof":          ["birth", "age", "जन्म"],
        "birth-certificate":  ["birth certificate", "janm"],
        "caste-certificate":  ["caste", "जाति"],
        "disability-certificate": ["disability", "दिव्यांग"],
        "bpl-certificate":    ["bpl", "below poverty"],
        "project-report":     ["project", "परियोजना"],
        "incorporation-cert": ["incorporation", "cin"],
        "business-plan":      ["business plan", "funding"],
        "maternal-card":      ["maternal", "anc", "गर्भ"],
        "sowing-certificate": ["sowing", "crop", "बुवाई"],
    }
    text = f"{file_name} {hint}".lower()
    for doc_type, keys in KEYWORD_MAP.items():
        if any(k in text for k in keys):
            return doc_type
    return "unknown"


def analyze_document(file_name: str = "", content: dict = {}, hint: str = "") -> dict:
    """Analyze document metadata for the JSON /upload endpoint."""
    doc_type = _keyword_classify(file_name, hint)
    fields   = DOCUMENT_TYPES.get(doc_type, {}).get("fields", [])
    extracted = {f: content.get(f, f"[extracted {f}]") for f in fields} if doc_type != "unknown" else {}
    return {
        "type":            doc_type,
        "label":           DOCUMENT_TYPES.get(doc_type, {}).get("label", "Unknown Document"),
        "confidence":      0.4 if doc_type == "unknown" else 0.92,
        "extractedFields": extracted,
        "warnings":        [],
        "fileName":        file_name or "uploaded-file",
        "engine":          "keyword-mock",
        "analyzedAt":      datetime.now(timezone.utc).isoformat(),
    }
